elastic/beats
# Avoid duplicate names in dynamic_templates

## What does this PR do?
When multiple dynamic templates are detected for the same field, add the
matching type to the template name for uniqueness. 

## Why is it important?
As described in https://github.com/elastic/beats/issues/17203#issuecomment-603436023 the APM Server depends on allowing multiple dynamic templates for the same field, with different matching types. Therefore this PR takes care of adding some unique part to the name rather than failing with an error. 


## Checklist



~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have made corresponding change to the default configuration files~

## How to test this PR locally
This can be tested with APM Server as it creates multiple dynamic templates for the `labels` field. 
* Check out this branch and APM Server master and redirect to use this branch instead of the remote beats version. Can be achieved by running `go mod edit -replace github.com/elastic/beats/v7=<path-to-your-local-beats-repo>`.
* start ES instance; compile and run APM Server `make update apm-server && ./apm-server -e -E setup.template.overwrite=true`
* fetch the apm-server template `apm-8.0.0` and search for `labels_string`, `labels_boolean` and `labels_*`. 




## Related issues


* closes elastic/beats#17203
* fixes https://github.com/elastic/apm-server/issues/2693


## Screenshots
<img width="394" alt="Screenshot 2020-05-29 at 15 54 04" src="https://user-images.githubusercontent.com/5555349/83267584-b7f26700-a1c4-11ea-84c5-239d94995372.png">
